File: src/code/java/structures/ListNode.java
package code.java.structures;

import java.util.ArrayList;
import java.util.List;

public class ListNode {
    public int val;
    public ListNode next;

    public ListNode() {
    }

    public ListNode(int val) {
        this.val = val;
    }

    public ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    public ListNode(List<Integer> list) {
        this(new ArrayList<>(list));
    }

    public ListNode(ArrayList<Integer> list) {
        this.val = list.removeFirst();
        if (!list.isEmpty())
            this.next = new ListNode(list);
    }

    @Override
    public String toString() {
        if (next != null)
            return val + ", " + next;
        else
            return Integer.toString(val);
    }
}

File: src/code/kotlin/structures/ListNode.kt
package code.kotlin.structures

class ListNode(var `val`: Int) {
    var next: ListNode? = null

    constructor(list: ArrayList<Int>) : this(list[0]) {
        if (list.size > 1)
            next = ListNode(list.subList(1, list.size))
    }

    constructor(list: Collection<Int>) : this(ArrayList(list))

    constructor(array: IntArray) : this(array.toList())

    constructor(`val`: Int, next: ListNode): this(`val`) {
        this.next = next
    }

    override fun toString(): String {
        return `val`.toString() + if (next != null) ", ${next.toString()}" else ""
    }
}